### DISCUSSION TITLE
Weird behavior with SSR

### DISCUSSION THREAD

**Discussion by thierrymichel** [vuejs/pinia#948] [Help and Questions] [answered]
Hi all,

I am super happy to make our stack evolve and move from `vuex` to `pinia` (as well as moving from `vue2` to `vue3` and  from `webpack` to `vite`) 😁.
Thanks for the great job! 🙏

Almost everything is perfect except one weird behaviour.
Nothing blocking but I would like to have your lights to understand… 😇

Thanks in advance,
Thierry

---

### Reproduction

Clone this repo : [pinia-vite-ssr-issue](https://github.com/thierrymichel/pinia-vite-ssr-issue) and use `yarn dev`

### Steps to reproduce the behavior

Check the logs and the browser.

### Expected behavior

Store instance should have 1 item (as initial state and active pinia instance do).

![image](https://user-images.githubusercontent.com/806883/148369852-440a5d55-89a8-4c4b-99bd-dda7cc6b7601.png)

![image](https://user-images.githubusercontent.com/806883/148369813-ff47b8a8-6dd6-4f0f-a58b-69578b2a3422.png)

### Actual behavior

Store instance is empty.

![image](https://user-images.githubusercontent.com/806883/148369917-126b8ee0-72dd-47db-b89e-5eca0d37b3d3.png)

### Additional information

I've added comments and snippets inside the `README` file.

**Comment by posva** [comment]
I’m glad you like it!
There is too much code for the reproduction, can you remove all the routing part and make the code smaller so I can check?

**Reply by posva** [reply]
I transferred this to a discussion in the meantime

**Reply by thierrymichel** [reply]
Ok. I will make the code smaller, removing the routing part 👍

**Reply by thierrymichel** [reply]
I made it as simple and readable as possible…
[https://github.com/thierrymichel/pinia-vite-ssr-issue/tree/main/src](https://github.com/thierrymichel/pinia-vite-ssr-issue/tree/main/src)

**Reply by posva** [reply]
Thanks! that looks a lot smaller but it doesn't run anything when visiting localhost 🤔 
It's also weird that calling `useStore()` inside of a sync function changes anything. Have you tried explicitly passing pinia to `useStore(pinia)`? Maybe that works out in your project

**Reply by thierrymichel** [reply]
Yes. Also tried, doesn't change anything, same results 😕

**Reply by posva** [reply]
Right now the repro doesn't run, it gives an empty response when visiting localhost (BTW you can remove everything linting related too, it's not used). Are you sure you are not missing a file somewhere?

**Reply by thierrymichel** [reply]
Mmmmh. Made a fresh install (`clone`, `yarn`, `yarn dev`) without any problem.
I removed config files and everything unneeded.

- I use NodeJS v16.13.0
- It's on `https`, port `3000`

Beside that 🧐

**Reply by posva** [reply]
I was having trouble with the https option, disabling it allowed me to run the repro :+1:

**Comment by posva** [comment, marked as answer]
Now I see the problem: on the client side you need to make sure you hydrate the state before using the store:

```js
    const pinia = createPinia()

    app.use(pinia)

    // This does not work
    if (import.meta.env.SSR) {
      doSomethingWithStore()
      const store = useStore()
      store.fetchSomething()
      initialState.pinia = pinia.state.value
    } else {
      pinia.state.value = initialState.pinia
      doSomethingWithStore()
    }
```

On server you **finish** by serializing the state to the client while on the client you **start by reading (hydrating) the state** before using the store.

**Reply by thierrymichel** [reply]
I not sure I got it.
The counter intuitive thing is that I do not really "do-something-with-store".
I "just" register vue-router guards that will use store actions later…

Calling `useStore()` seems to generate another "state instance" (?), different from the active one ([`getActivePinia`](https://github.com/thierrymichel/pinia-vite-ssr-issue/blob/main/src/App.vue#L14)).

I've updated the repo to reflect that ([https://github.com/thierrymichel/pinia-vite-ssr-issue/blob/main/src/module.ts](https://github.com/thierrymichel/pinia-vite-ssr-issue/blob/main/src/module.ts)).

Thanks for your time.

**Reply by posva** [reply]
any `useStore()` call in client side should go **after hydration**.

Let me know if something else isn't clear?

**Comment by niklv** [comment]
Found this behaviour in my Nuxt 4 project.
Only workaround that i found: 
Call all of stores that should by hydreted in my plugin. Before any usage in vue components.
